### PR TITLE
Added validation for type name uniqueness in a schema

### PR DESCRIPTION
**Pull request by bartekbp**
## Description
Today, it's possible to unintentionally define a schema that uses the same type name twice. You can have two modules with the following:
```python
# query.py
@strawberry.type
class User:
    name: str


@strawberry.type
class Query:
    viewer: User
    management: Management
```

```python
# management.py
@strawberry.type
class User:
    firstname: str
    lastname: str
    created_at: int

@strawberry.type
class Management:
    users: list[User]
```

The resulting schema will contain only one type to represent a user, thus leading to a conflict - one type will shadow the other one.

This pr introduces a validation ensuring that if type with the same name is duplicated, its definition must be the same.

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [x] Enhancement/optimization
- [ ] Documentation

## Issues Fixed or Closed by This PR

* #404 

## Checklist

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
- [x] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by botberry** [comment]
Thanks for adding the `RELEASE.md` file!



Here's a preview of the changelog:

---

Added extra validation that types used in a schema are unique.
Strawberry starts to throw an exception `DuplicatedTypeName` when two types defined in a schema have the same name.


---
Here's the preview release card for twitter:




Here's the tweet text: 
```
🆕 Release (next) is out! Thanks to Bartosz Polnik for the PR 👏

Get it here 👉 https://github.com/strawberry-graphql/strawberry/releases/tag/(next)
```

File: strawberry/schema/schema_converter.py
Comment (DoctorJohn) [inline strawberry/schema/schema_converter.py:673]: We probably don't want to keep those prints. If the information they're printing is relevant they should be part of the exception :) What do you think?

```suggestion
```

File: strawberry/exceptions.py
Comment (DoctorJohn) [inline strawberry/exceptions.py:251]: @strawberry-graphql/core This is not specific to this PR, but I'm wondering whether we eventually want all our errors to extend a common base exception, so that strawberry errors are easier identifiable and catch-able? (or is this even considered an anti-pattern, since users would be tempted to catch the base strawberry exception instead of more specific ones?)

**Review by DoctorJohn — CHANGES_REQUESTED** [review]
Thanks for the PR @bartekbp! Overall I like these changes, makes strawberry even more polished. I added one suggestion and annoyed the whole core team by inviting them to discuss something not even specific to this PR lol

File: strawberry/exceptions.py
Comment (patrick91) [reply strawberry/exceptions.py:251]: yes, this is the plan, I think we can leave this PR as it is, and I'll update #2027 when this one is merged :)

Comment (DoctorJohn) [reply strawberry/exceptions.py:251]: Ah, I see. Thanks Patrick!

File: strawberry/schema/schema_converter.py
Comment (bartekbp) [reply strawberry/schema/schema_converter.py]: Indeed, this was just for debugging. I'll remove it:)

**Review by DoctorJohn — APPROVED** [review]
LGTM :)
(except for the repeating pattern, but there's probably not much we can do about it)

**Review by patrick91 — APPROVED** [review]
This is great! Thank so much! <3

**Comment by patrick91** [comment]
@bartekbp thanks for contributing to Strawberry! I'll merge this PR once it's green 🎉 Also, I've invited you to join the Strawberry GraphQL organisation 😊

You can also request a free sticker by filling this form: https://forms.gle/dmnfQUPoY5gZbVT67

And don't forget to join our discord server: https://strawberry.rocks/discord 🔥

**Pull request merged**
